projects/Mallard/src/services/logging.ts

```diff
@@ -0,0 +1,190 @@
+import {
+    iapReceiptCache,
+    userDataCache,
+    loggingQueueCache,
+} from 'src/helpers/storage'
+import { errorService } from './errors'
+import { getCASCode } from 'src/authentication/helpers'
+import Config from 'react-native-config'
+import {
+    Level,
+    Feature,
+    MallardLogFormat,
+} from '../../../Apps/common/src/logging'
+
+const { LOGGING_API_KEY } = Config
+
+interface LogParams {
+    level: Level
+    message: string
+    optionalFields?: object
+}
+
+class Logging {
+    async getExternalInfo() {
+        const [
+            networkStatus,
+            userData,
+            casCode,
+            iapReceipt,
+        ] = await Promise.all([
+            NetInfo.fetch(),
+            userDataCache.get(),
+            getCASCode(),
+            iapReceiptCache.get(),
+        ])
+        return {
+            networkStatus,
+            userData,
+            casCode,
+            iapReceipt,
+        }
+    }
+
+    async baseLog({
+        level,
+        message,
+        ...optionalFields
+    }: LogParams): Promise<MallardLogFormat> {
+        const {
+            networkStatus,
+            userData,
+            casCode,
+            iapReceipt,
+        } = await this.getExternalInfo()
+
+        // User Data and Subscription
+        const userId =
+            (userData && userData.userDetails && userData.userDetails.id) || ''
+        const digitalSub =
+            (userData &&
+                userData.membershipData &&
+                userData.membershipData.contentAccess &&
+                userData.membershipData.contentAccess.digitalPack) ||
+            false
+        const iAP = iapReceipt ? true : false
+
+        return {
+            app: DeviceInfo.getBundleId(),
+            version: DeviceInfo.getVersion(),
+            buildNumber: DeviceInfo.getBuildNumber(),
+            os: Platform.OS === 'ios' ? 'ios' : 'android',
+            device: DeviceInfo.getDeviceId(),
+            networkStatus: networkStatus
+                ? networkStatus.type
+                : NetInfoStateType.unknown,
+            release_channel: isInBeta() ? 'BETA' : 'RELEASE',
+            timestamp: new Date(),
+            level,
+            message,
+            deviceId: DeviceInfo.getUniqueId(),
+            signedIn: userData ? true : false,
+            userId,
+            digitalSub,
+            casCode,
+            iAP,
+            ...optionalFields,
+        }
+    }
+
+    async getQueuedLogs(): Promise<MallardLogFormat[] | [{}]> {
+        try {
+            const logString = await loggingQueueCache.get()
+            return JSON.parse(logString || '[{}]')
+        } catch (e) {
+            return [{}]
+        }
+    }
+
+    async saveQueuedLogs(log: MallardLogFormat[]): Promise<string | Error> {
+        try {
+            const logString = JSON.stringify(log)
+            await loggingQueueCache.set(logString)
+            return 'saved logs'
```

Comment:
does this return string actually get used anywhere?

Reply:
No it doesnt. More just for completeness. If you were to use this function and wanted some positive response that everything was fine, you can at least receive a string. `loggingQueueCache` returns `void` so there is no "feedback"